### DISCUSSION TITLE
How to enable GPU mode while creating an index?

### DISCUSSION THREAD

**Discussion by xiaojuan8** [milvus-io/milvus#15655] [Q&A and General discussion]
Hi, I have started the milvus-server docker with command "docker run -d --name milvus_gpu_1.1.1 --gpus device=2 \
-p 19530:19530 \
-p 19121:19121 \
-v /data0/usr/xiaojuan8/milvus/db:/var/lib/milvus/db \
-v /data0/usr/xiaojuan8/milvus/conf:/var/lib/milvus/conf \
-v /data0/usr/xiaojuan8/milvus/logs:/var/lib/milvus/logs \
-v /data0/usr/xiaojuan8/milvus/wal:/var/lib/milvus/wal \
milvusdb/milvus:1.1.1-gpu-d061621-330cc6"

and I inserted 4 billion embedding vector into a collection, and I used an index "HNSW(M=32, efConstruction=128)" ,when I called creating an index to milvus-server, nvidia-smi shows no usage of graphics card(gpu2) as shown below:

![image](https://user-images.githubusercontent.com/74544828/154884750-1db3c720-c265-4520-8969-080f63b392ad.png)

**Comment by xiaojuan8** [comment]
Is there any config that I missed while creating an index ?  help me with any info, thanks

**Comment by yhmo** [comment]
Not all index types can support GPU mode. In v1.x, only IVF series index can use GPU device to build/search. Your index type is HNSW which only supports CPU mode.

**Reply by xiaojuan8** [reply]
OK，btw， Is there any config that I missed when I created the index? Just start the milvus-server in docker with gpu image， I can request to create an IVF series index with params(nlist .etc). Do I need to config any params with gpu info in request to milvus-server?

**Reply by yhmo** [reply]
Just make sure the **server_config.yaml** gpu mode is enabled before you launch milvus server:
```
gpu:
  enable: true
  cache.enable: false
  cache_size: 1GB
  gpu_search_threshold: 1000
  search_devices:
    - gpu0
  build_index_devices:
    - gpu0
```

- **gpu.enable**, set to true to enable GPU mode
- **gpu.cache.enable**, if set to true, milvus will cache segment data in GPU memory(limited by the gpu.cache_size)
- **gpu.cache_size**, set this value to be less than your GPU memory size, typically 1/2 ~ 2/3 of the GPU memory size
- **gpu.gpu_search_threshold**, this is for search(), if the nq(number of query target vectors) of search() is larger than this value, milvus will try use GPU to search. If you always want GPU search, you can set this value to be 0 or 1.
- **gpu.search_devices**, specify which GPU device is working for search(), it can be multiple devices, just use comma to split multiple devices like this:  gpu0, gpu2
   When a search request is received, milvus will dispatch segments to different GPU devices to compute. But if there is only one segment in the collection, you will see only one GPU device is busy, others are idle.

- **gpu.build_index_devices**, specify which GPU device is working for building index, it can be multiple devices, just use comma to split multiple devices like this:  gpu1, gpu3

The **create_index()** and **search()** has no parameter for GPU, use same parameters for CPU mode and GPU mode.

**Reply by xiaojuan8** [reply]
👌thanks.

**Reply by Yufei447** [reply]
I only found milvus.yaml, i don't have server_config.yaml. I tried add the gpu section in the milvus.yaml, but it seems that it still use cpu.

**Reply by yhmo** [reply]
> I only found milvus.yaml, i don't have server_config.yaml. I tried add the gpu section in the milvus.yaml, but it seems that it still use cpu.


Because server_config.yaml is the configuration file of Milvus v1.0. The v1.0 can support GPU computation.

**Reply by Escapethebug** [reply]
I now have mivlus 2.3.0 installed, how can I configure it so that milvus applies GPU for both indexing and querying. appreciated!

**Reply by yhmo** [reply]
> I now have mivlus 2.3.0 installed, how can I configure it so that milvus applies GPU for both indexing and querying. appreciated!

CPU Index type: IVF_FLAT
GPU index type: GPU_IVF_FLAT

If you create GPU_IVF_FLAT, both indexing and querying are in GPU.
If you carate IVF_FLAT, both indexing and querying are in CPU.

The official v2.3.0 is coming soon.

**Reply by Escapethebug** [reply]
Thank you very much for your reply! I have installed 2.3.0beta and there is no GPU_IVF_FLAT, is it possible that all indexes are built and queried on the CPU. Is there a way to use the GPU, or are we waiting for the official version 2.3.0 . As just using milvus not very familiar with these, thank you again!

**Reply by yhmo** [reply]
Just wait for the v2.3.0. Next Tuesday.

**Reply by Escapethebug** [reply]
Thank you! Looking forward to it.

**Reply by Escapethebug** [reply]
It's still a beta version on the official website, how long will it take for the official 2.3.0 version to be released? Thank you very much!

**Reply by yhmo** [reply]
We got block issues and the release is denied. Will release it in one or two days.

**Reply by pancodia** [reply]
Is it released? 
I followed [this milvus doc](https://milvus.io/docs/install_standalone-gpu-docker.md) to deploy the stack. 
Then, I am using pymivlus Version: 2.3.0 to create collection and index with the following index params. However, I didn't see any process used GPU. 
Is there a way that I can test if the stack is properly deployed and GPU is used?

```
index_params = {
    'metric_type': "IP",
    'index_type': 'GPU_IVF_FLAT', #"IVF_FLAT",
    'params':{
        "nlist": 128
    }
}
```

**Reply by pancodia** [reply]
My bad. After creating the index and insert some data, when I load the collection, I can see the GPU is being used as the memory occupied is increasing. 

Is there any parameters I need to pass during search if I want to make sure GPU is used during search time?
